Tests/Functional/Drivers/Queue/PhpAmqpLibDriverTest.php

```diff
@@ -0,0 +1,259 @@
+    public function testSend(MessageInterface $msg)
+    {
+        $msgIn = $this->createQueueMessage($msg);
+        $msgIn->addHeader('test_header', '12345');
+        $this->assertTrue($this->driver->send($msgIn));
+    }
+
+    /**
+     * Test the consume process without a callback in the return
+     */
+    public function testConsumeWithoutCallback()
+    {
+        $this->consumer = $this->createConsumer();
+        $consumerTag = $this->consumer->getName();
+        $this->driver->declareChannel();
+        $this->driver->declareQueue($this->queueName, QueueMessage::DELIVERY_MODE_PERSISTENT, []);
+        $return = $this->driver->consume($this->consumer->getName(), $this->queueName);
+        $this->assertNull($return);
+    }
+
+    /**
+     * Tests the consume process with a callback in return that receives the messages and ack this after
+     *
+     * @dataProvider getMessages
+     */
+    public function testConsumeWithCallbackAckingMessage(MessageInterface $msg)
+    {
+        $this->prepareToConsume($msg);
+
+        $amqpMessage = null;
+        $callback = function(AMQPMessage $message) use (&$amqpMessage){
+            $amqpMessage = $message;
+            $queueMessage = new QueueMessage();
+            $queueMessage->setMessageId($amqpMessage->getDeliveryTag());
+            $this->driver->ack($queueMessage);
+        };
+
+        $this->driver->consume($this->consumer->getName(), $this->queueName, $callback);
+        $this->driver->declareChannel()->is_consuming();
+        $this->driver->wait();
+
+        $this->assertInstanceOf(AMQPMessage::class, $amqpMessage);
+        $this->assertEquals($amqpMessage->delivery_info['routing_key'], $this->queueName);
+        $this->assertEquals($amqpMessage->delivery_info['consumer_tag'], $this->consumer->getName());
+        $this->assertNotNull($amqpMessage->getBody());
+    }
+
+    /**
+     * Tests the consume process with a callback in return that receives the messages and nack this after
+     *
+     * @dataProvider getMessages
+     */
+    public function testConsumeWithCallbackNackingMessage(MessageInterface $msg)
+    {
+        $this->prepareToConsume($msg);
+        $amqpMessage = null;
+
+        $callback = function($message) use (&$amqpMessage) {
+            $amqpMessage =  $message;
+            $queueMessage = new QueueMessage();
+            $queueMessage->setMessageId($amqpMessage->getDeliveryTag());
+            $this->driver->nack($queueMessage);
+        };
+
+        $this->driver->consume($this->consumer->getName(), $this->queueName, $callback);
+        $this->driver->declareChannel()->is_consuming();
+        $this->driver->wait();
+
+        $this->assertInstanceOf(AMQPMessage::class, $amqpMessage);
+        $this->assertEquals($amqpMessage->delivery_info['routing_key'], $this->queueName);
+        $this->assertEquals($amqpMessage->delivery_info['consumer_tag'], $this->consumer->getName());
+        $this->assertNotNull($amqpMessage->getBody());
```

Comment:
The only way to check a nack is counting the messages in the queue before and after the consume. For this test here it will be complicated once to get this information I have to call the function declareQueue that already is called inside the consume function, what turns it in a hackish way and duplicated calls.

Reply:
If a message is nacked (as we have implemented the code) the message comes back to the queue from where it came and as a header that message has the reason of the rejection. This is the "x-death" header that is an array and inside there is field call **reason** that in the case of the nack the value is **rejected**

Reply:
We can improve this in the future